reactor-aeron-core/src/main/java/reactor/aeron/MessagePublication.java

```diff
@@ -43,130 +49,123 @@
     this.connectTimeout = options.connectTimeout();
     this.backpressureTimeout = options.backpressureTimeout();
     this.adminActionTimeout = options.adminActionTimeout();
-    this.writeLimit = options.resources().writeLimit();
   }
 
   /**
-   * Enqueues content for future sending.
+   * Enqueues publisher for future sending.
    *
-   * @param buffer abstract buffer to send
+   * @param publisher abstract publisher to process messages from
    * @param bufferHandler abstract buffer handler
    * @return mono handle
    */
-  <B> Mono<Void> publish(B buffer, DirectBufferHandler<? super B> bufferHandler) {
+  <B> Mono<Void> publish(Publisher<B> publisher, DirectBufferHandler<? super B> bufferHandler) {
     return Mono.defer(
         () -> {
-          if (isDisposed()) {
-            return Mono.error(AeronExceptions.failWithPublicationUnavailable());
-          }
-          MonoProcessor<Void> promise = MonoProcessor.create();
-          if (!publishTasks.offer(
-              PublishTask.newInstance(publication, buffer, bufferHandler, promise))) {
-            return Mono.error(AeronExceptions.failWithPublicationUnavailable());
-          }
-          return promise;
+          PublisherProcessor processor = new PublisherProcessor(bufferHandler, this);
+          publisher.subscribe(processor);
+          return processor.onDispose();
```

Comment:
we should add double checks, first just to fail fast and another to correctly close/remove/release resources

```
          if (isDisposed()) {
            return Mono.error(AeronExceptions.failWithPublicationUnavailable());
          }
```
and add at end double

Reply:
Putting `if` s is not best way to fix race conditions.

Reply:
I propose to create and issue in reactor aeron and assign this issue on me. In a meantime let's move on with relaese of the new version.

Reply:
@artem-v ok but we must have at least 1 check for it. Otherwise it won't work at all